File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        // Base cases
        if (subRoot == null) {
            return true;  // Empty tree is always a subtree
        }
        if (root == null) {
            return false;  // Can't find subtree in an empty tree
        }
        
        // Check if the trees match from current root
        if (isSameTree(root, subRoot)) {
            return true;
        }
        
        // Recursively check left and right subtrees
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }
    
    // Helper method to check if two trees are identical
    private boolean isSameTree(TreeNode p, TreeNode q) {
        // If both are null, they're the same
        if (p == null && q == null) {
            return true;
        }
        
        // If one is null and the other isn't, they're different
        if (p == null || q == null) {
            return false;
        }
        
        // Check if values are the same and recursively check subtrees
        return p.val == q.val && 
               isSameTree(p.left, q.left) && 
               isSameTree(p.right, q.right);
    }
}

File: Data Structures & Algorithms/same-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        self.same_tree = True
        self.dfs(p, q)
        return self.same_tree

    def dfs(self, p, q):
        # If one is null and the other is not, trees differ
        if p is None or q is None:
            if p != q:
                self.same_tree = False
            return
        
        # If values differ, trees differ
        if p.val != q.val:
            self.same_tree = False
            return

        self.dfs(p.left, q.left)
        self.dfs(p.right, q.right)